src/main/scala/com/gu/fezziwig/CirceScroogeWhiteboxMacros.scala

```diff
@@ -0,0 +1,301 @@
+      } else {
+        q"${param.typeSignature}"
+      }
+      tq"""_root_.shapeless.::[_root_.shapeless.labelled.FieldType[${witnessName}.T, ${fieldType}], $acc]"""
+    }
+    val hlist = params.foldRight[Tree](q"""_root_.shapeless.HNil""") { case (param, acc) =>
+      val paramName = TermName(param.name.toString())
+      q"""_root_.shapeless.::(${paramName}, $acc)"""
+    }
+    // hlistPattern is identical to hlist but with the pq interpolator instead of q
+    val hlistPattern = params.foldRight[Tree](pq"""_root_.shapeless.HNil""") { case (param, acc) =>
+      val paramName = TermName(param.name.toString())
+      pq"""_root_.shapeless.::(${paramName}, $acc)"""
+    }
+    val labelledFields = params.map(param => {
+      val paramName = TermName(param.name.toString())
+      val witnessName = TermName(s"${param.name.toString()}Witness")
+      val (fieldExpr, fieldType): (Tree, Tree) = if (param.asTerm.isParamWithDefault) {
+        (q"_root_.scala.Some(struct.${paramName})", tq"Option[${param.typeSignature}]")
+      } else {
+        (q"struct.${paramName}", q"${param.typeSignature}")
+      }
+      q"""val ${paramName}: _root_.shapeless.labelled.FieldType[${witnessName}.T, ${fieldType}] = _root_.shapeless.labelled.field(${fieldExpr})"""
+      }
+    )
+    val labelledGeneric = q"""
+    new _root_.shapeless.LabelledGeneric[$A] {
+      type Repr = ${reprType}
+
+      def to(struct: $A): Repr = {
+        ..$labelledFields
+        $hlist
+      }
+
+      def from(hlist: Repr): $A = hlist match {
+        case $hlistPattern => $apply(..${fieldNames})
+      }
+    }"""
+    q"""{
+      ..$witnesses
+
+      $labelledGeneric
+    }"""
+  }
+
+  /**
+    * Provides a LabelledGeneric implicit for Thrift unions, as
+    * [[thriftStructGeneric]] does for structs.
+    *
+    * Given this thrift definition
+    *
+    * {{{
+    * union Union1 {
+    *   1: StructC c
+    *   2: StructD d
+    * }
+    * }}}
+    *
+    * the produced implicit will be roughly equivalent to
+    *
+    * {{{
+    * {
+    *   import shapeless.{Witness, :+:, labelled, CNil, Inl, Inr}
+    *
+    *   val dWitness = Witness(Symbol("d"))
+    *   val cWitness = Witness(Symbol("c"))
+    *   type Union1Repr = labelled.FieldType[dWitness.T, StructD] :+: labelled.FieldType[cWitness.T, StructC] :+: CNil
+    *
+    *   new LabelledGeneric[Union1] {
+    *     type Repr = Union1Repr
+    *
+    *     def to(union: Union1): Repr = union match {
+    *       case Union1.D(x) => Inl(labelled.field(x))
+    *       case Union1.C(x) => Inr(Inl(labelled.field(x)))
+    *       case Union1.UnknownUnionField(_) => throw new RuntimeException("Unknown union field")
+    *     }
+    *
+    *     def from(repr: Repr): Union1 = repr match {
+    *       case Inl(d) => Union1.D(d)
+    *       case Inr(Inl(c)) => Union1.C(c)
+    *       case Inr(Inr((_: CNil))) => throw new RuntimeException("Encountered CNil while converting from ReprType")
```

Comment:
I wonder if an incomplete match would be better.

Reply:
I’ve realised that taking out the `: CNil` type signature removes the warning, so I’m gonna go with that to avoid putting annoying warnings in consumers’ code.